File: 5-longest-palindromic-substring/longest-palindromic-substring.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        if s == s[::-1]:
            return s
        
        start, max_len = 0, 1

        for i in range(1, len(s)):
            l, r = i-max_len, i+1
            s1 = s[l-1:r] # +2 check
            s2 = s[l:r] # +1 check

            if l >= 1 and s1 == s1[::-1]:
                start = l-1
                max_len += 2
            elif s2 == s2[::-1]:
                start = l
                max_len += 1
        
        return s[start: start+max_len]

File: 5-longest-palindromic-substring/longest-palindromic-substring.ts
function longestPalindrome(s: string): string {
    if (s.length <= 1) { return s; }

    let start = 0;
    let maxLength = 1;

    const expandAroundCenter = (left: number, right: number): number => {
        while (left >= 0 && right < s.length && s[left] === s[right]) {
            left--;
            right++;
        }
        return right - left - 1;
    }

    for (let i = 0; i < s.length; i++) {
        let len1 = expandAroundCenter(i, i);
        let len2 = expandAroundCenter(i, i+1);
        let currMaxLength = Math.max(len1, len2);
        if (currMaxLength > maxLength) {
            maxLength = currMaxLength;
            start = i - Math.floor((maxLength - 1) / 2);
        }
    }
    return s.substring(start, start + maxLength);
};